src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexFacetedReadOperation.cs

```diff
@@ -0,0 +1,423 @@
+                        facetValues[iterator.IsNull ? Constants.NullValue : Constants.EmptyString].IncrementCount(1);
+                        continue;
+                    }
+                    InsertTerm(iterator.Sequence, ref entryReader);
+                }
+        
+                break;
+            case IndexEntryFieldType.Tuple:
+            case IndexEntryFieldType.Simple:
+                if (fieldReader.Read(out var source))
+                    InsertTerm(source, ref entryReader);
+                break;
+            case IndexEntryFieldType.Null:
+                break;
+            default:
+                throw new Exception($"Got type {fieldReader.Type}");
+        }
+        
+         void InsertTerm(ReadOnlySpan<byte> source, ref IndexEntryReader entryReader)
+         {
+             var nameAsSlice = GetFieldNameAsSlice(result.Value.AggregateBy);
+             analyzersScope.Execute(nameAsSlice, source, out var buffer, out var tokens);
+
+             foreach (var tokenOutput in tokens)
+             {
+                 token.ThrowIfCancellationRequested();
+                 var term = buffer.Slice(tokenOutput.Offset, (int)tokenOutput.Length);
+                 var encodedTerm = Encodings.Utf8.GetString(term);
+
+
+                 if (facetValues.TryGetValue(encodedTerm, out var collectionOfFacetValues) == false)
+                 {
+                     var range = FacetedQueryHelper.GetRangeName(result.Value.AggregateBy, encodedTerm);
+                     collectionOfFacetValues = new FacetValues(legacy);
+                     if (needToApplyAggregation == false)
+                         collectionOfFacetValues.AddDefault(range);
+                     else
+                     {
+                         foreach (var aggregation in result.Value.Aggregations)
+                             collectionOfFacetValues.Add(aggregation.Key, range);
+                     }
+
+                     facetValues.Add(encodedTerm, collectionOfFacetValues);
+                 }
+                 
+                 collectionOfFacetValues.IncrementCount(1);
+
+                 if (needToApplyAggregation)
+                 {
+                     ApplyAggregation(result.Value.Aggregations, collectionOfFacetValues, ref entryReader);
+                 }
+             }
+         }
+    }
+
+    private void GetFieldReader(ref IndexEntryReader reader, in string name, out IndexEntryReader.FieldReader fieldReader)
+    {
+        if (_fieldMappings.TryGetByFieldName(name, out var binding))
+        {
+            // In this case we've to check if field is dynamic also
+            fieldReader = reader.GetReaderFor(binding.FieldId);
+            return;
+        }
+
+        var slicedFieldName = GetFieldNameAsSlice(name);
+        fieldReader = reader.GetReaderFor(slicedFieldName);
+    }
+
+    private void ApplyAggregation(Dictionary<FacetAggregationField, FacetedQueryParser.FacetResult.Aggregation> aggregations, FacetValues values,
```

Comment:
The common logic here is not complicated. Making some generic method for that would be pretty hard (because we process data differently in Lucene/Corax)

Reply:
ok